docs-site/docs/dist-train.md

```diff
@@ -0,0 +1,432 @@
+
+SINGA supports distributed data parallel training and evaulation process based on multiprocessing. The following is the illustration of the data parallel training:
+
+![MPI.png](assets/MPI.png)
+
+In the distributed training, each process runs a training script which utilizes one GPU. Each process has an individual rank, which gives information of which GPU the individual process is using. The training data is partitioned, so that each process can evaluate the sub-gradient based on the partitioned training data. Once the sub-graident is calculated on each processes, the overall stochastic gradient is obtained by all-reducing the sub-gradients evaluated by all processes. The all-reduce operation is supported by the NVidia Collective Communication Library (NCCL).
+
+The all-reduce operation by NCCL can be used to reduce and synchronize the parameters from different GPUs. Let's consider a data partitioned distributed training using 4 GPUs. Once the sub-gradients from the 4 GPUs are calculated, the NCCL can perform the all-reduce process so that all the GPUs can get the sum of the sub-gradients over the GPUs:
+
+![AllReduce.png](assets/AllReduce.png)
+
+Finally, the parameter update of Stochastic Gradient Descent (SGD) can then be performed by using the overall stochastic gradient obtained by the all-reduce process.
+
+## Python DistOpt Methods:
+
+There are a list of methods for distributed training with DistOpt:
+
+1. Create a DistOpt with the SGD object and device assignment:
+
+```python
+sgd = opt.SGD(lr=0.005, momentum=0.9, weight_decay=1e-5)
+sgd = opt.DistOpt(sgd)
+dev = device.create_cuda_gpu_on(sgd.rank_in_local)
+```
+
+&nbsp;
+
+2. Backward propagation and distributed parameter update:
+
+```python
+sgd.backward_and_update(loss)
+```
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;loss is the objective function of the deep learning model optimization,
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;e.g. for classification problem it can be the output of the softmax_cross_entropy function.
+
+&nbsp;
+
+3. Backward propagation and distributed parameter update, using half precision for gradient communication:
+
+```python
+sgd.backward_and_update_half(loss)
+```
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;It converts the gradients to 16 bits half precision format before allreduce
```

Comment:
@chrishkchris Hi, sorry for troubling you again, but may I know if you need these `&nbsp;` in these markdown files? thanks

Reply:
yes, those &nbsp create an empty space/line for better formatting when we view it on browser
thanks

Reply:
got it. Thanks for the clarification!